xtm/xtm_api.cc

```diff
@@ -0,0 +1,355 @@
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * <COPYRIGHT HOLDER> OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+#include "xtm_api.h"
+#include "xtm_scsp_queue.h"
+#include "xtm_config.h"
+
+#include <stdint.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <assert.h>
+#include <fcntl.h>
+#include <errno.h>
+#ifdef TARANTOOL_XTM_USE_EVENTFD
+#include <sys/eventfd.h>
+#endif /* defined(TARANTOOL_XTM_USE_EVENTFD) */
+
+#define XTM_PIPE_SIZE 4096
+
+union xtm_msg {
+	/**
+	 * Anonymous structure for storing functions and their arguments,
+	 * when using the xtm API in dispatch/invoke pattern.
+	 */
+	struct {
+		xtm_queue_fun_t fun;
+		void *fun_arg;
+	};
+	/**
+	 * Pointer to save message, when using xtm API
+	 * in send/recv pattern.
+	 */
+	void *ptr;
+};
+
+struct xtm_queue {
+	/**
+	 * File descriptor that the consumer thread must poll,
+	 * to know when new messages are added to the queue.
+	 */
+	int consumer_read_fd;
+	/**
+	 * File descriptor to which the producer thread writes
+	 * to notify the consumer thread of new messages in the queue.
+	 */
+	int consumer_write_fd;
+	/**
+	 * File descriptor that the producer thread must poll,
+	 * to know when there is a free space in qeuue.
+	 */
+	int producer_read_fd;
+	/**
+	 * File descriptor to which the consumer thread writes
+	 * to notify the producer thread about free space in queue.
+	 */
+	int producer_write_fd;
+	/**
+	 * Flag indicates, that producer, wants to get notification
+	 * about free place in queue.
+	 */
+	bool is_producer_need_notifications;
+	/**
+	 * Flag indicates, that xtm_delete must close all read file
+	 * descriptors, otherwise they must be closed by user before
+	 * calling xtm_delete.
+	 */
+	bool is_need_to_close_readfd;
+	/**
+	 * Flag indicates, that producer couldn't put an item in
+	 * the queue, and wait for notification.
+	 */
+	bool is_notification_expected;
+	/** Message queue, it's size must be power of two */
+	struct xtm_scsp_queue<union xtm_msg> queue;
+};
+
+static inline int
+notify_fd(int fd)
+{
+	static uint64_t tmp = 1;
+	ssize_t cnt;
+	/*
+	 * We must write 8 byte value, because for linux we
+	 * used eventfd, which require to write >= 8 byte at once.
+	 * Also in case of EINTR we retry to write
+	 */
+	while ((cnt = write(fd, &tmp, sizeof(tmp))) < 0 && errno == EINTR)
+		;
+	return (cnt == sizeof(tmp) ? 0 : -1);
```

Comment:
I just realized. We probably should not consider it an error when notify fails due to the pipe being full. Because it means it is already notified. Can it be detected somehow? I don't remember. All the other errors are rare, should not happen ever. But this one might happen. And it is not really a problem, we could ignore the error.
What does libev do?

Reply:
libev, ignore errors, it's not check return code of `write` function:
`uint64_t counter = 1;`
`write (evpipe [1], &counter, sizeof (uint64_t));`
But i don't think that it is good idea for us. If write fails with errno == EAGAIN, that means, that pipe is full, we can ignore it (although it's stange that reader did not have time to flush pipe). But other errors mean something went wrong and user must know about it!

Reply:
I am thinking about cases when a user does not want to support any unlikely errors, and he just wants to make `panic()` in case of any error. And then we really might want to ignore EAGAIN, treat it like a success. Because it means the consumer is already notified anyway. Yeah, other errors must not be ignored of course.